File: python/백준/백트래킹/15683 감시.py
import sys
from itertools import product

input = sys.stdin.readline

N, M = map(int, input().split())

board = []
cctv = []

for i in range(N):
    row = list(map(int, input().split()))
    board.append(row)
    for j in range(M):
        if 1 <= row[j] <= 5:
            cctv.append((i, j, row[j]))

types = {
    1: [[0], [1], [2], [3]],
    2: [[0, 2], [1, 3]],
    3: [[0,1], [1, 2], [2, 3], [3, 0]],
    4: [[0, 1, 2], [1, 2, 3], [2, 3, 0], [3, 0 ,1]],
    5: [[0, 1, 2, 3]]
}

dx = [-1, 0, 1, 0]
dy = [0, 1, 0 ,-1]

def watch(b, r, c, d):
    while True:
        r, c = r + dx[d], c + dy[d]
        if r < 0 or r >= N or c < 0 or c >= M:
            break
        if b[r][c] == 6:
            break
        if b[r][c] == 0:
            b[r][c] = '#'

choices = [range(len(types[t])) for _, _, t in cctv]

answer = N * M

for combo in product(*choices):
    b = [row[:] for row in board]
    for idx, choice in enumerate(combo):
        r, c, t = cctv[idx]
        for d in types[t][choice]:  
            watch(b, r, c, d)
    count = sum(row.count(0) for row in b)
    answer = min(answer, count)

print(answer)

File: cpp/백준/백트래킹/15683 감시.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <tuple>
using namespace std;

int N, M;
vector<vector<int>> board;
vector<tuple<int, int, int>> cctv;

vector<vector<vector<int>>> types = {
    {},
    {{0}, {1}, {2}, {3}},
    {{0, 2}, {1, 3}},
    {{0, 1}, {1, 2}, {2, 3}, {3, 0}},
    {{0, 1, 2}, {1, 2, 3}, {2, 3, 0}, {3, 0, 1}},
    {{0, 1, 2, 3}}
};

int dx[] = {-1, 0, 1, 0};
int dy[] = {0, 1, 0, -1};

void watch(vector<vector<int>>& b, int r, int c, int d) {
    while (true) {
        r += dx[d];
        c += dy[d];
        if (r < 0 || r >= N || c < 0 || c >= M) break;
        if (b[r][c] == 6) break;
        if (b[r][c] == 0) b[r][c] = -1;
    }
}

int answer;

void solve(int idx, vector<vector<int>>& b) {
    if (idx == (int)cctv.size()) {
        int count = 0;
        for (int i = 0; i < N; ++i)
            for (int j = 0; j < M; ++j)
                if (b[i][j] == 0) count++;
        answer = min(answer, count);
        return;
    }

    int r, c, t;
    tie(r, c, t) = cctv[idx];

    for (int choice = 0; choice < (int)types[t].size(); ++choice) {
        vector<vector<int>> copy = b;
        for (int d : types[t][choice]) {
            watch(copy, r, c, d);
        }
        solve(idx + 1, copy);
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M;
    board.assign(N, vector<int>(M));
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < M; ++j) {
            cin >> board[i][j];
            if (board[i][j] >= 1 && board[i][j] <= 5) {
                cctv.push_back({i, j, board[i][j]});
            }
        }
    }

    answer = N * M;
    solve(0, board);
    cout << answer << "\n";
    return 0;
}
